chalk-solve/src/clauses.rs

```diff
@@ -18,6 +18,49 @@ mod env_elaborator;
 mod generalize;
 pub mod program_clauses;
 
+// yields the types "contained" in `app_ty`
+fn constituent_types<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    app_ty: &ApplicationTy<I>,
+) -> Vec<Ty<I>> {
```

Comment:
It would be nice to not collect into a Vec here, but I think that can be left for future optimizations.

Reply:
Well, yield would work nicely if it'd be stable; or one could construct an enum (implementing Iterator) to dispatch between the two possible iterators.
Its fun doing that ^^
But I think the performance gain would be outweighed by the loss of readability.
What do you think?

Reply:
Or just construct the clauses in this function! But yeah, for later.